paddle/fluid/framework/program_processing.cc

```diff
@@ -0,0 +1,139 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/framework/program_processing.h"
+#include "paddle/fluid/framework/block_desc.h"
+
+namespace paddle {
+namespace framework {
+
+void ProgramProcessor::GetInputsOutputsInBlock(const BlockDesc &current_block,
+                                               VariableNameMap *inner_inputs,
+                                               VariableNameMap *inner_outputs) {
+  /* Find inputs and outputs in current control flow block.
+  :param current_block: Current control flow block.
+  :param inner_inputs: Input VariableNameMap of ops in current block.
+  :param inner_outputs: Output VariableNameMap of ops in current block. */
+
+  // Step1: update inner_inputs and inner_outputs
+  // NOTE: Here assumes that all variables are input or output of Ops,
+
+  std::set<std::string> inner_inputs_var;
+  std::set<std::string> inner_outputs_var;
+
+  for (OpDesc *op : current_block.AllOps()) {
+    for (auto iname : op->InputNames()) {
+      bool AddInput = false;
+      for (auto in_var_name : op->Input(iname)) {
+        VLOG(3) << "in_var_name:" << in_var_name;
+        if (inner_outputs_var.find(in_var_name) == inner_outputs_var.end()) {
+          inner_inputs_var.insert(in_var_name);
+          AddInput = true;  // here we assume if one var is not in the
+                            // outputs_var list, we add this input into the
+                            // inputs list.
+        }
+      }
+      if (AddInput) inner_inputs->emplace(iname, op->Input(iname));
```

Comment:
Since we only care about variable names, not op input parameter names like 'X', 'Y', I suggest you can use `std::set<std::string>` instead of the `VariableNameMap`

Reply:
Done.